Update docs on automatic stats refresh rate

rmloveland opened the pull request:
... and node restarts after stats deletion to clear caches.

Summary of changes:

- Add a new subsection to CBO page, 'Controlling statistics refresh
  rate', where we describe the cases when stats are refreshed in more
  detail.

- To match the structure of the above, we break the instructions for
  deleting stats into a new section 'Turning off statistics'

- Finally, tweak stats deletion instructions on CBO page and CREATE
  STATS page so both say that nodes must be restarted
  post-stats-deletion to clear caches.

Fixes #4809, #4872.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/4925)
<!-- Reviewable:end -->

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ee681b7af7efdec317e83593fe3feb37e402c2e2/

Edited pages:
- [v19.1/cost-based-optimizer.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ee681b7af7efdec317e83593fe3feb37e402c2e2/v19.1/cost-based-optimizer.html)
- [v19.2/cost-based-optimizer.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ee681b7af7efdec317e83593fe3feb37e402c2e2/v19.2/cost-based-optimizer.html)

rmloveland commented:
@rytaft Thanks for the details.  I tried to split the difference language-wise so we're not too constrained if the implementation of refresh rate needs to change.  PTAL and let me know what you think.  I am happy to add more detail if you think that's too vague.  I guess my sense is that if users want to mess with these settings they should really know what they're doing (at least that's what the docs say currently) so I'm loathe to document exact formulas. But I defer to your expertise!

@RaduBerinde I snuck in the update to fix #4872 (stats deletion & node restart to clear cache) as well since it's in the same area, mind taking a look?

rytaft commented:
Thanks for this update, @rmloveland! I'm not too worried about the implementation of the refresh rate changing, since in that case we'll need to change those cluster settings and will update the docs anyway (although we could change the defaults). Regardless, I do think this is a good level of detail. The one thing I'd suggest is to add a sentence or two about how changing the settings qualitatively affects the refresh rate. For example, you could mention that increasing either `min_stale_rows` or `fraction_stale_rows` reduces the frequency of refreshes. `min_stale_rows` in particular impacts the frequency of refreshes for small tables, while `fraction_stale_rows` has more of an impact on larger tables.

Maybe @RaduBerinde also has an opinion about the level of detail...





<!-- Sent from Reviewable.io -->

RaduBerinde commented on v19.1/cost-based-optimizer.md:
```diff
@@ -67,10 +67,20 @@ The cost-based optimizer can often find more performant query plans if it has ac
+
+| Setting                                              | Details                                                                              |
+|------------------------------------------------------+--------------------------------------------------------------------------------------|
+| `sql.stats.automatic_collection.fraction_stale_rows` | Target fraction of stale rows per table that will trigger a statistics refresh       |
```
Do we want to include the default values here? Seems like it would be useful, but there's the downside that they could become stale. Not sure what the general practice is for cluster settings.

rmloveland replied:
Our general practice for cluster settings (thus far) is to try really hard to point users to the (generated, up-to-date) cluster settings page, unless we really need to mention the settings.  Since we need them here in this case, I'd rather not include the default values here for the reason you said: staleness / the risk of the values being incorrect.

rmloveland commented:
Thanks for the reviews @rytaft and @RaduBerinde !  I hope my latest updates incorporate your feedback correctly, please let me know if you think further updates are needed.

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ed3431e919864215998326e4f1375527148c5f80/

Edited pages:
- [v19.1/cost-based-optimizer.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ed3431e919864215998326e4f1375527148c5f80/v19.1/cost-based-optimizer.html)
- [v19.2/cost-based-optimizer.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ed3431e919864215998326e4f1375527148c5f80/v19.2/cost-based-optimizer.html)

rytaft commented:
Looks good, @rmloveland! I think you might just want to add a bit to this sentence:
> you should not expect to see statistics update immediately after changing these settings

In addition to expecting a refresh after changing the settings, users might also expect a refresh after a certain number of rows have changed. For example, If a table is empty, users might expect a refresh after exactly 500 rows have been inserted (the default for `min_stale_rows`). But the refresh could actually happen before or after that.

rmloveland commented:
rytaft <notifications@github.com> writes:

> Looks good, @rmloveland! I think you might just want to add a bit to
> this sentence:
>
>     you should not expect to see statistics update immediately after
>     changing these settings
>
> In addition to expecting a refresh after changing the settings, users
> might also expect a refresh after a certain number of rows have
> changed. For example, If a table is empty, users might expect a
> refresh after exactly 500 rows have been inserted (the default for
> min_stale_rows). But the refresh could actually happen before or after
> that.

Thanks Becca!  Updated to note that 500 rows will not necessarily
trigger the refresh.  In order for that sentence to even make sense, I
went ahead and added the default values for each setting to this page
instead of just relying on the cluster settings page (so Radu I guess I
lied before, sorry. :-} ).

PTAL

cockroach-teamcity commented:
Online preview: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9e305bdea01ada9117b29fbae22873e0842bd721/

Edited pages:
- [v19.1/cost-based-optimizer.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9e305bdea01ada9117b29fbae22873e0842bd721/v19.1/cost-based-optimizer.html)
- [v19.2/cost-based-optimizer.md](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9e305bdea01ada9117b29fbae22873e0842bd721/v19.2/cost-based-optimizer.html)

rytaft commented:
LGTM thanks!
